### ISSUE TITLE
ValidationScope: User unfriendly implementation

### ISSUE THREAD

**Issue by juniwalk** [nette/forms#111]
Hello,
I'd like to ask, could we come up with some way to rework disabling validation scope for selected fields?

Currently if you have 17 fields and one of them is AJAXified `<select>`, you have to do this:
```php
$form->addSubmit('submit')->setValidationScope([
        $form['name'], $form['email'], $form['password'], $form['phone'], $form['ic'],
        $manage['isBanned'], $manage['isApproved'], $account, $address,
]);
```
Best way would be to disable this per-field.

**Comment by dg** [comment]
What about condition?

```php
$form->addSelect('country', 'Country:', $countries);
$form->addSubmit('submit', 'Send');

$form['country']
	->addConditionOn($form['submit'], $form::SUBMITTED)
		->setRequired('Choose country');
```

**Comment by juniwalk** [comment]
How does this help me when I have those countries loaded using AJAX and have `$select->checkAllowedValues = FALSE;`? I need to disable scope validation for that select for the submit to work.

Sorry if I haven't made myself clear about that.

**Comment by dg** [comment]
Why you need to disable scope validation for select?

**Comment by juniwalk** [comment]
Because I am stupid, `$select->setPrompt()` is what I needed, not to disable validation scope. But I still think that disablind the scope per field would be better.

**Comment by juniwalk** [comment]
@dg I am closing this as it is probably unnecessary to have it here.

**Comment by juniwalk** [comment]
@dg I have to disable validation scope when I use `$select->setRequired();`, `$select->setPrompt()` is not applicable in this case.

The problem is when I submit form with AJAX data, the value of select is outside `$data` variable, I have to get it using `$form->getHttpData()` which is okay but the form fails to submit because value of select is required and there is none in the time of checking.

I am back to my initial issue, disabling validation scope in this way is not friendly.

**Comment by dg** [comment]
Can you try to implement any solution?

**Comment by juniwalk** [comment]
@dg Well I have a few ideas

### First idea
- Add `$validateScope` property to `BaseControl` which will be bool. 
- Allow disabling scope per control using `$control->disableValidationScope();` // just examples
- Before the form is builded, iterate over each control and build validationScope and assign it to SubmitButton
- **But** what if there is more than one `SubmitButton` control? Is that possible? That would cause problems.

### Second idea
- Add class `Nette\Forms\ValidationScopeBuilder`
- Implement `\Traversable` or extend `\ArrayObject`
- Pass form instance into it
- Call `$vsb->disableScopeFor('control');`
- Pass instance into `$submit->setValidationScope($vsb);`
- This will get all current controls from the form, remove those disabled and foreach in `setValidationScope` will iterate over them.

### Third idea
- Each BaseControl can get it's parent Form instance
- We could add disableValidationScopeFor('control') on the SubmitButton control
- Add those control names into list and then just select all controls from form and leave out disabled ones

What do you think? Does any of those sound good?

**Comment by JanTvrdik** [comment]
> the form fails to submit because value of select is required and there is none in the time of checking.

Why not just set `$select->checkAllowedValues = FALSE`? Redesigning validation scope seems like overkill.

**Comment by juniwalk** [comment]
@JanTvrdik Yeah, it is particularly troubling issue this one. I don't mind working with validation scope, it's just not very intuitive to work with.

**Comment by JanTvrdik** [comment]
I don't think that validation scopes should be used for this kind of issue at all.

Another random idea

~~~php
$form->addSelect('country', 'Country', [$form->getHttpData('country') => '...'])
    ->setRequired();
~~~

**Comment by juniwalk** [comment]
> @JanTvrdik I don't think that validation scopes should be used for this kind of issue at all.

Really? Why?

With your another random idea, I see potential problem with setting default value for editing.

**Comment by JanTvrdik** [comment]
@juniwalk Because if you don't want the value to be validated, you should not make the select box required.

**Comment by juniwalk** [comment]
@JanTvrdik I do want the value validated, and I need it to be required, but that does not play well with AJAX values feeded in to the select box ...

**Comment by JanTvrdik** [comment]
> I do want the value validated

Could you elaborate on what exactly and where do you validate?

**Comment by juniwalk** [comment]
@JanTvrdik In this context I ment the requirement validation as i use this to suggest user names and I get back just int (user Id) which I then in the onSuccess (probably not properly done but that is how I use it).

**Comment by JanTvrdik** [comment]
I'm sorry, I still don't understand it. How could you use the server-side validation when the select box is not validated at all because it is not included in the validation scope? What did I miss?

**Comment by juniwalk** [comment]
@JanTvrdik Ignore it, we diverged from the problem here, validation is not the issue, getting the value is.

**Comment by JanTvrdik** [comment]
Getting the value is not an issue if the value is not validated https://github.com/nette/forms/issues/111#issuecomment-191263297

**Comment by juniwalk** [comment]
@JanTvrdik Yes, and to disable the scope is absolute horrendous work to do (especially for large forms).

**Comment by dg** [comment]
`setValidationScope(array_diff_key($form->getControls(), ['country' => TRUE]))` is horrendous?

**Comment by juniwalk** [comment]
@dg Hey, now that looks much better, would you mind if I send PR to docs with this example to help others like me?

**Comment by dg** [comment]
Of course I'll be happy. But try it please, I wrote from my head.

**Comment by juniwalk** [comment]
@dg So I tried that array_diff_key and it throws an error
`PHP Warning: array_diff_key(): Argument #1 is not an array in ./app/forms/ProductForm.php:136`

**Comment by juniwalk** [comment]
Hi,
I revisited this old issue. Now when I am more skilled in Nette I basically never modify validationScope, I do my own validation in onValidate event.

I was modifying it before to keep the clientSide validation so user gets alerted before submitting that he did not select something required, however I can live without it.

I think we can close this issue.
